softhsm: init -> 2.1.0

leenaars opened the pull request:
SoftHSM is an implementation of a cryptographic store accessible through a PKCS #11 interface. You can use it to explore PKCS #11 without having a Hardware Security Module. It is being developed as a part of the OpenDNSSEC project. SoftHSM uses Botan for its cryptographic operations.

exi commented on pkgs/tools/security/softhsm/default.nix:
```diff
@@ -0,0 +1,30 @@
+  meta = {
+    homepage = https://www.opendnssec.org/softhsm/;
+    description = "cryptographic store accessible through a PKCS #11 interface";
+    license = stdenv.lib.licenses.bsd;
```
There is no stdenv.lib.licenses.bsd license

exi commented on pkgs/tools/security/softhsm/default.nix:
```diff
@@ -0,0 +1,30 @@
+    "--localstatedir=$out/var"
+    ];
+
+  buildInputs = [ botan];
```
Small nitpick: Please add a space before the ]

exi commented on pkgs/tools/security/softhsm/default.nix:
```diff
@@ -0,0 +1,30 @@
+  configureFlags = [
+    "--with-crypto-backend=botan"
+    "--with-botan=${botan}"
+    "--sysconfdir=$out/etc"
```
Would this not make configuration impossible? $out/etc is mounted readonly usually.

exi commented on pkgs/tools/security/softhsm/default.nix:
```diff
@@ -0,0 +1,30 @@
+{ stdenv, fetchurl, unzip, botan }:
```
unzip is never used

exi commented:
@leenaars Why close the pull request? More packages are always welcome. If my tone was too strict, i sincerely apologize.

leenaars commented:
Hi there, Github automatically closed it when I send a forced updated of my branch on the server. I was trying to fix the commit while looking at the output of the continuous build, so I somehow missed your comments. I will resubmit, because it is really useful software.

exi commented:
@leenaars strange, i always force-push my branches and github never closes it.

leenaars commented:
I don't know what triggered it. I guess the fact that I removed the commits referenced in the pull request caused it. The whole juggling of branches has quite a learning curve anyway...

I've fixed the space before the ], I'll research your comment on the /etc. Any suggestion?

exi commented:
@leenaars Thanks. Sadly i'm not that familiar with that part myself (not enough to know how it should be done properly). Best course of action would be to ask in the nixos irc channel.
@ip1981 @devhell @globin any ideas?

leenaars commented:
Searching through the repo it seems there are many packages that similarly use $out/etc. Any suggestions very welcome.

ip1981 commented:
It usually requires runtime configuration by systemd services.

leenaars commented:
Softhsm is intended to be used mainly as a library [1], I'm told it in the typical usage scenario it is not actually necessary to configure it with systemd or the likes - if any configuration is needed at all, system variables will do. I've moved softhsm to development/libraries. 

I've also removed the unnecessary dependency on unzip, and checked that it still builds.

[1] https://wiki.opendnssec.org/display/SoftHSMDOCS/SoftHSM+Documentation+v2.0

fpletz commented on pkgs/top-level/all-packages.nix:
```diff
@@ -3090,7 +3090,7 @@ let
 
   snort = callPackage ../applications/networking/ids/snort { };
-
+  
```
Please remove that trailing whitespace. :)

leenaars commented:
I have rebased the patch, and removed the extra spaces.

cillianderoiste commented on pkgs/development/libraries/softhsm/default.nix:
```diff
@@ -0,0 +1,30 @@
+    "--with-crypto-backend=botan"
+    "--with-botan=${botan}"
+    "--sysconfdir=$out/etc"
+    "--localstatedir=$out/var"
```
Does it expect to write things to $out/var ... because that would fail since it's not possible to write to the store. I'm not sure what the correct path would be, is there a suitable default?

zimbatm replied:
Usually we just use /var in the other recipes

zimbatm commented:
@leenaars almost there ! Can you rebase and squash your commits ?

leenaars commented:
@zimbatm: I've rebased and squashed, and pointed localstatedir to /var. Anything else I should do?

zimbatm commented:
@leenaars /var/lib/softhsm/tokens is now part of the package but I suppose it needs to be writable right ?

leenaars commented:
@zimbatm: you are right, that is exactly where I got stuck. As you can see, I tried creating /var/lib/softhsm/tokens during preInstall, but even as root that action is not allowed. When localstatedir is set to /var, install-data-hook fails with "No such file or directory" ...

fpletz commented:
It looks like you will have to patch the Makefile so it doesn't create these files/directories, or if they are needed let the Makefile install them into `$out` somewhere and make sure softhsm can find it.

zimbatm commented:
@leenaars either patch the Makefile or remove it after install: `postInstall = "rm -rf $out/var";`. The issue is that the /nix/store is mounter read-only and you are never supposed to write to it (only the nix-daemon can do such thing).
After install if the binary complains that /var/lib/softhsm/tokens doesn't exist, you could make a wrapper that `mkdir -p /var/lib/softhsm/tokens` before executing the program.

leenaars committed:
softhsm: init -> 2.1.0

leenaars commented:
Meanwhile a new release has been made, v. 2.1.0. I've included that release with the suggestion from @zimbatm. Apologies for the long delay.

zimbatm commented:
:+1: it looks like some of the earlier fixes has been lost so I fixed them and pushed 38c5319fd8a9c01ef70d27bc98970bdc230451cc to master. Thanks !
